### PR TITLE
[REF] Cleanup validation in PaypalProIPN single function

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
[REF] Cleanup validation in PaypalProIPN single function

Before
----------------------------------------
Passed in object used for validation. Properties set on the passed-in-object only to be discarded

After
----------------------------------------
Simplified validation, unused variables not set

Technical Details
----------------------------------------
This fixes interaction with the contribution object to instead use the function
and switches the standardised error handling (throw an exception & let
it be caught).

The contribution object used to be passed to completeOrder but no longer
so there is no reason to set total_amount & invoice_id on it
(they will be used from input)

Comments
----------------------------------------

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@seamuslee001 are you OK to merge this - it's actually pretty straight forward

File: CRM/Core/Payment/PayPalProIPN.php [hunk 1]
```diff
@@ -364,26 +364,16 @@ public function single($input, $contribution, $recur = FALSE, $first = FALSE) {
...
-      if ($contribution->invoice_id != $input['invoice']) {
-        Civi::log()->debug('PayPalProIPN: Invoice values dont match between database and IPN request.');
-        echo "Failure: Invoice values dont match between database and IPN request<p>contribution is" . $contribution->invoice_id . " and input is " . $input['invoice'];
-        return;
+      if ($this->getContributionObject()->invoice_id !== $input['invoice']) {
+        throw new CRM_Core_Exception('PayPalProIPN: Invoice values dont match between database and IPN request.');
       }
     }
-    else {
-      $contribution->invoice_id = md5(uniqid(rand(), TRUE));
```
Comment (seamuslee001) [inline CRM/Core/Payment/PayPalProIPN.php:374]: @eileenmcnaughton do we not need to do this for subsequent recurring contributions?

File: CRM/Core/Payment/PayPalProIPN.php [hunk 2]
```diff
@@ -364,26 +364,16 @@ public function single($input, $contribution, $recur = FALSE, $first = FALSE) {
 
     // make sure the invoice is valid and matches what we have in the contribution record
     if ((!$recur) || ($recur && $first)) {
-      if ($contribution->invoice_id != $input['invoice']) {
-        Civi::log()->debug('PayPalProIPN: Invoice values dont match between database and IPN request.');
-        echo "Failure: Invoice values dont match between database and IPN request<p>contribution is" . $contribution->invoice_id . " and input is " . $input['invoice'];
-        return;
+      if ($this->getContributionObject()->invoice_id !== $input['invoice']) {
```
Comment (seamuslee001) [inline CRM/Core/Payment/PayPalProIPN.php:367]: This simplification is fine to me

File: CRM/Core/Payment/PayPalProIPN.php [hunk 3]
```diff
@@ -364,26 +364,16 @@ public function single($input, $contribution, $recur = FALSE, $first = FALSE) {
...
-    else {
-      $contribution->invoice_id = md5(uniqid(rand(), TRUE));
-    }
 
     if (!$recur) {
-      if ($contribution->total_amount != $input['amount']) {
-        Civi::log()->debug('PayPalProIPN: Amount values dont match between database and IPN request.');
-        echo "Failure: Amount values dont match between database and IPN request<p>";
-        return;
+      if ($this->getContributionObject()->total_amount != $input['amount']) {
```
Comment (seamuslee001) [inline CRM/Core/Payment/PayPalProIPN.php:373]: This simplification is fine to me

File: CRM/Core/Payment/PayPalProIPN.php [hunk 4]
```diff
@@ -364,26 +364,16 @@ public function single($input, $contribution, $recur = FALSE, $first = FALSE) {
...
+        throw new CRM_Core_Exception('PayPalProIPN: Amount values dont match between database and IPN request.');
       }
     }
-    else {
```
Comment (seamuslee001) [inline CRM/Core/Payment/PayPalProIPN.php:384]: This works for me as well

File: CRM/Core/Payment/PayPalProIPN.php [hunk 5]
```diff
@@ -402,7 +392,7 @@ public function single($input, $contribution, $recur = FALSE, $first = FALSE) {
       Contribution::update(FALSE)->setValues([
         'cancel_date' => 'now',
         'contribution_status_id:name' => 'Cancelled',
-      ])->addWhere('id', '=', $contribution->id)->execute();
+      ])->addWhere('id', '=', $this->getContributionID())->execute();
```
Comment (seamuslee001) [inline CRM/Core/Payment/PayPalProIPN.php:395]: Yep fine

File: CRM/Core/Payment/PayPalProIPN.php [hunk 6]
```diff
@@ -419,7 +409,7 @@ public function single($input, $contribution, $recur = FALSE, $first = FALSE) {
       return;
     }
 
-    CRM_Contribute_BAO_Contribution::completeOrder($input, $this->getContributionRecurID(), $contribution->id ?? NULL);
+    CRM_Contribute_BAO_Contribution::completeOrder($input, $this->getContributionRecurID(), $this->getContributionID());
```
Comment (seamuslee001) [inline CRM/Core/Payment/PayPalProIPN.php:412]: Fine

File: CRM/Core/Payment/PayPalProIPN.php (hunk 1 above)
Comment (eileenmcnaughton) [reply CRM/Core/Payment/PayPalProIPN.php:374]: @seamuslee001 no - I did try to explain in the message template - that contribution doesn't actually 'go anywhere' from here - or at least nowhere that accesses this property

Comment (seamuslee001) [reply CRM/Core/Payment/PayPalProIPN.php:374]: right i see now ok

File: CRM/Core/Payment/PayPalProIPN.php (hunk 6 above)
Comment (eileenmcnaughton) [reply CRM/Core/Payment/PayPalProIPN.php:412]: See how repeat other never gets the contribution - so that invoice_id is set but discarded

**Comment by seamuslee001** [comment]
All seems fine to me now

**Pull request merged**